src/cookbook/gradual-compositionapi-migration.md

```diff
@@ -0,0 +1,258 @@
+
+If you're using Vuex, it's very clear exactly what centralized state is being used across the application. In addition, you gain the benefit of using a standardized pattern that comes with officially supported tooling in the DevTools which can make a big difference when it comes to debugging code.
+
+Composition API is very flexible, but you may lose that implicit declaration in communication to fellow maintainers. Our suggestion that if you do use it as a centralized state management store, that you place it in a `stateManagement` folder, or something similarly named, so that responsibilities are clear.
+
+## Components
+
+It is possible at this point to start to move your components to the Composition API, however, for larger code bases, you may not see immediate need or good reason to, as the Options API works nicely.
+
+Our suggestion is to use this opportunity to think through the architecture and refactor any larger components that have too much responsibility on their own.
+
+For instance, if you have a component has the following methods:
+
+- one that gathers input from a form
+- one that validates that input
+- one that uses the input to call out to an API
+- one that does something with that API response
+
+You may find that you can potentially create a reusable component for the API call, as on a large application there's a fairly high chance you will need to make another.
+
+For the purposes of demonstration, here's how that one method refactor might look.
+
+### Before, with a Tightly Coupled Method
+
+```js
+methods: {
+  async getGitHubProjects() {
+    try {
+      this.isLoading = true
+      let projects = await fetch(
+        `https://api.github.com/users/sdras/repos?page=1&per_page=100`
+      ).then(res => res.json());
+
+      this.gitHubProjects = projects
+    } catch (error) {
+      this.error = error
+      console.error(error);
+    } finally {
+      this.isLoading = false;
+    }
+  }
+},
+created() {
+  this.getGitHubProjects()
+}
+```
+
+### After, with Composition API
+
+```js
+import { reactive, toRefs } from '@vue/composition-api'
+
+export function useFetchAPI(api, options) {
+  const state = reactive({
+    response: null,
+    isLoading: false,
+    hasErrors: null
+  })
+
+  const apiCall = async () => {
+    state.isLoading = true
+    try {
+      let res = await fetch(api, options)
+      state.response = await res.json()
+    } catch (error) {
+      state.hasErrors = error
+    } finally {
+      state.isLoading = false
+    }
+  }
+
+  return {
+    ...toRefs(state),
+    apiCall
+  }
+}
+```
+
+And now we can refactor our earlier component like this:
+
+```js
+import { useFetchAPI } from '@/composables/useFetchAPI.js'
+
+export default {
+  data() {
+    return {
+      gitHubData: [],
+      errors: null,
+      loading: false
+    }
+  },
+  methods: {
+    async callGitHub() {
```

Comment:
I don't think there's any need for the `async` here. The function doesn't return anything and it isn't using `await`.

Shouldn't this example still have the `created` hook?

More generally, I find this example a little odd. I guess I'd have expected most of this logic to be in `setup` with `callGitHub` just being the returned alias for `apiCall`. That might just be my lack of experience with the Composition API...

Reply:
It's not on the created hook because the example is called from a button. While it's possible to call the api with created and then display with a method, that's a lot more for someone to parse while trying to understand something else. The rest is dealt with in another comment, but thanks!